modules/server/src/main/scala/scaladex/server/service/UserSessionSynchronizer.scala

```diff
@@ -0,0 +1,73 @@
+package scaladex.server.service
+
+import java.util.UUID
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+import akka.actor.ActorSystem
+import cats.implicits.toTraverseOps
+import scaladex.core.model.UserInfo
+import scaladex.core.model.UserState
+import scaladex.core.service.GithubService
+import scaladex.core.service.SchedulerDatabase
+import scaladex.infra.GithubClient
+
+class UserSessionSynchronizer(database: SchedulerDatabase)(implicit ec: ExecutionContext, ac: ActorSystem)
+  extends Scheduler("user-session-synchronizer", 1.hour) {
+
+  override def run(): Future[Unit] =
+    for {
+      sessions <- database.getAllSessions()
+      sessionsWithClients = createClientsForEach(sessions)
+      updatedMaybeUserInfos <- sessionsWithClients.traverse {
+        case (userId, session) => getUserInfo(userId, session)
+      }
+      sessionsToDeleteIds = expiredSessionIds(sessions, updatedMaybeUserInfos)
+      sessionsToUpdate = updatedSessions(sessions, updatedMaybeUserInfos)
+      _ <- sessionsToUpdate.traverse {
+        case (userId, userState) => database.insertSession(userId, userState)
+      }
+      _ <- sessionsToDeleteIds.traverse(database.deleteSession)
+    } yield ()
+
+  private def createClientsForEach(sessions: Seq[(UUID, UserState)]): Seq[(UUID, GithubService)] =
+    sessions.map {
+      case (userId, userState) => (userId, new GithubClient(userState.info.token))
+    }
+
+  private def getUserInfo(userId: UUID, service: GithubService): Future[Option[(UUID, UserInfo)]] =
+    service.getUserInfo().map(userInfo => Option.apply((userId, userInfo)))
+      .recoverWith {
+        case _ =>
+          logger.info(s"Token for user with id: '$userId' is likely expired")
```

Comment:
Not sure what the retention policy is like for logs, would this be GDPR-compliant? We're already storing the `userID` as a primary key in the database anyway, just wanted to be sure.

Reply:
We keep the logs for only one month. Not sure about GDPR compliance but I would say that a expired token is not a personal piece of information.